exchanges/gateio/gateio_types.go

```diff
@@ -25,526 +54,2550 @@ var (
+	currency.BUY:      100,
+	currency.INSUR:    290,
+	currency.PUSH:     79,
 }
 
-// OrderbookResponse stores the orderbook data
-type OrderbookResponse struct {
-	Result  string `json:"result"`
-	Elapsed string `json:"elapsed"`
-	Asks    [][]string
-	Bids    [][]string
+// CurrencyInfo represents currency details with permission.
+type CurrencyInfo struct {
+	Currency         string `json:"currency"`
+	Delisted         bool   `json:"delisted"`
+	WithdrawDisabled bool   `json:"withdraw_disabled"`
+	WithdrawDelayed  bool   `json:"withdraw_delayed"`
+	DepositDisabled  bool   `json:"deposit_disabled"`
+	TradeDisabled    bool   `json:"trade_disabled"`
+	FixedFeeRate     string `json:"fixed_rate,omitempty"`
+	Chain            string `json:"chain"`
+}
+
+// CurrencyPairDetail represents a single currency pair detail.
+type CurrencyPairDetail struct {
+	ID              string  `json:"id"`
+	Base            string  `json:"base"`
+	Quote           string  `json:"quote"`
+	Fee             float64 `json:"fee"`
+	MinBaseAmount   float64 `json:"min_base_amount"`
+	MinQuoteAmount  float64 `json:"min_quote_amount"`
+	AmountPrecision int     `json:"amount_precision"`
+	Precision       int     `json:"precision"`
+	TradeStatus     string  `json:"trade_status"`
+	SellStart       int     `json:"sell_start"`
+	BuyStart        int     `json:"buy_start"`
+}
+
+// Ticker holds detail ticker information for a currency pair
+type Ticker struct {
+	CurrencyPair     string    `json:"currency_pair"`
+	Last             float64   `json:"last"`
+	LowestAsk        float64   `json:"lowest_ask"`
+	HighestBid       float64   `json:"highest_bid"`
+	ChangePercentage string    `json:"change_percentage"`
+	ChangeUtc0       string    `json:"change_utc0"`
+	ChangeUtc8       string    `json:"change_utc8"`
+	BaseVolume       float64   `json:"base_volume"`
+	QuoteVolume      float64   `json:"quote_volume"`
+	High24H          float64   `json:"high_24h"`
+	Low24H           float64   `json:"low_24h"`
+	EtfNetValue      string    `json:"etf_net_value"`
+	EtfPreNetValue   string    `json:"etf_pre_net_value"`
+	EtfPreTimestamp  time.Time `json:"etf_pre_timestamp"`
+	EtfLeverage      float64   `json:"etf_leverage"`
+}
+
+// OrderbookData holds orderbook ask and bid datas.
+type OrderbookData struct {
+	ID      int         `json:"id"`
+	Current time.Time   `json:"current"` // The timestamp of the response data being generated (in milliseconds)
+	Update  time.Time   `json:"update"`  // The timestamp of when the orderbook last changed (in milliseconds)
+	Asks    [][2]string `json:"asks"`
+	Bids    [][2]string `json:"bids"`
+}
+
+// MakeOrderbook parse Orderbook asks/bids Price and Amount and create an Orderbook Instance with asks and bids data in []OrderbookItem.
+func (a *OrderbookData) MakeOrderbook() (*Orderbook, error) {
+	ob := &Orderbook{
+		ID:      a.ID,
+		Current: a.Current,
+		Update:  a.Update,
+	}
+	asks := make([]OrderbookItem, len(a.Asks))
+	bids := make([]OrderbookItem, len(a.Bids))
```

Comment:
as previously mentioned hug with bids for loops and then asks -> ob.Asks and for bids.